crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,504 @@
+use pico::Index;
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use crate::operation_text::hash;
+use common_lang_types::{
+    ArtifactFileName, ArtifactHash, ArtifactPathAndContent, FileContent, FileSystemOperation,
+    SelectableName, ServerObjectEntityName,
+};
+use isograph_config::PersistedDocumentsHashAlgorithm;
+
+#[derive(Debug, Clone, Default)]
+pub struct FileSystemState {
+    root_files: HashMap<ArtifactFileName, (Index<FileContent>, ArtifactHash)>,
+    nested_files: HashMap<
+        ServerObjectEntityName,
+        HashMap<SelectableName, HashMap<ArtifactFileName, (Index<FileContent>, ArtifactHash)>>,
+    >,
+}
+
+impl FileSystemState {
+    pub fn recreate_all(state: &Self, artifact_directory: &PathBuf) -> Vec<FileSystemOperation> {
+        let mut operations: Vec<FileSystemOperation> = Vec::new();
+        operations.push(FileSystemOperation::DeleteDirectory(
+            artifact_directory.to_path_buf(),
+        ));
+
+        for (server_object, selectables) in &state.nested_files {
+            let server_object_path = artifact_directory.join(server_object);
+
+            for (selectable, files) in selectables {
+                let selectable_path = server_object_path.join(selectable);
+                operations.push(FileSystemOperation::CreateDirectory(
+                    selectable_path.clone(),
+                ));
+
+                for (file_name, (index, _)) in files {
+                    let file_path = selectable_path.join(file_name);
+                    operations.push(FileSystemOperation::WriteFile(file_path, index.clone()));
+                }
+            }
+        }
+
+        for (file_name, (index, _)) in &state.root_files {
+            let file_path = artifact_directory.join(file_name);
+            operations.push(FileSystemOperation::WriteFile(file_path, index.clone()));
+        }
+
+        operations
+    }
+
+    pub fn diff(old: &Self, new: &Self, artifact_directory: &PathBuf) -> Vec<FileSystemOperation> {
+        let mut operations: Vec<FileSystemOperation> = Vec::new();
+
+        let mut new_server_objects: HashSet<ServerObjectEntityName> = HashSet::new();
+        let mut new_selectables: HashSet<(ServerObjectEntityName, &SelectableName)> =
```

Comment:
`SelectableName` instead of `&SelectableName`, since `SelectableName` implements `Copy` (and is small)